File: python/src/normalizer.py
"""Data normalizer for standard schema."""
import re
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field
from datetime import datetime

class JobPosting(BaseModel):
    """Normalized JobPosting schema."""
    title: str
    company: Optional[str] = None
    description: Optional[str] = None
    url: Optional[str] = None
    location: Optional[str] = None
    remote: bool = False
    job_type: Optional[str] = None
    salary_min: Optional[int] = None
    salary_max: Optional[int] = None

def normalize_jobposting(data: Dict[str, Any], source_url: str = "") -> Optional[JobPosting]:
    """Normalize JobPosting data to standard schema."""
    if not data.get('title'):
        return None
    
    # Extract company name
    company = None
    if 'hiringOrganization' in data:
        company = data['hiringOrganization'].get('name')
    elif 'company' in data:
        company = data['company']
    
    # Extract URL
    url = data.get('url') or data.get('application', {}).get('url') or source_url
    
    # Extract location
    location = None
    if 'jobLocation' in data:
        location = data['jobLocation'].get('address', {}).get('addressLocality')
    elif 'location' in data:
        location = data['location']
    
    # Detect remote
    remote = detect_remote(location or '')
    
    # Parse salary
    salary_min = None
    salary_max = None
    if 'estimatedSalary' in data:
        salary_min = parse_salary(data['estimatedSalary'].get('minValue'))
        salary_max = parse_salary(data['estimatedSalary'].get('maxValue'))
    
    return JobPosting(
        title=data.get('title', ''),
        company=company,
        description=data.get('description') or data.get('jobDescription'),
        url=url,
        location=location,
        remote=remote,
        job_type=data.get('employmentType'),
        salary_min=salary_min,
        salary_max=salary_max,
    )

def detect_remote(location_or_value: str) -> bool:
    """Detect if job is remote."""
    remote_keywords = ['remote', 'home', 'anywhere', 'worldwide', 'híbrido', 'hybrid']
    value = str(location_or_value).lower()
    return any(kw in value for kw in remote_keywords)

def parse_salary(value: Any) -> Optional[int]:
    """Parse salary to integer."""
    if not value:
        return None
    if isinstance(value, (int, float)):
        return int(value)
    if isinstance(value, str):
        cleaned = re.sub(r'[^\d]', '', value)
        return int(cleaned) if cleaned else None
    return None

def normalize_batch(raw_data: List[Dict], source_url: str = "") -> List[JobPosting]:
    """Normalize a batch of raw job data."""
    normalized = []
    for item in raw_data:
        try:
            job = normalize_jobposting(item, source_url)
            if job:
                normalized.append(job)
        except Exception:
            continue
    return normalized

File: node/src/scraper/normalizer.ts
interface JobPosting {
  title: string;
  company?: string;
  description?: string;
  url?: string;
  location?: string;
  remote: boolean;
  job_type?: string;
  salary_min?: number;
  salary_max?: number;
}

interface RawData {
  title?: string;
  url?: string;
  hiringOrganization?: { name?: string };
  company?: string;
  application?: { url?: string };
  jobLocation?: unknown;
  location?: string;
  estimatedSalary?: { minValue?: number | string; maxValue?: number | string };
  description?: string;
  jobDescription?: string;
  employmentType?: string;
}

const REMOTE_KEYWORDS = ['remote', 'home', 'anywhere', 'worldwide', 'híbrido', 'hybrid'];

function detectRemote(locationOrValue: string): boolean {
  const value = locationOrValue.toLowerCase();
  return REMOTE_KEYWORDS.some(kw => value.includes(kw));
}

function parseSalary(value: number | string | undefined): number | undefined {
  if (!value) return undefined;
  if (typeof value === 'number') return value;
  const cleaned = value.toString().replace(/\D/g, '');
  return cleaned ? parseInt(cleaned, 10) : undefined;
}

export function normalizeJobPosting(data: RawData, sourceUrl: string): JobPosting | null {
  if (!data.title) return null;

  // Extract company name
  let company: string | undefined;
  if (data.hiringOrganization?.name) {
    company = data.hiringOrganization.name;
  } else if (data.company) {
    company = data.company;
  }

  // Extract URL
  let url = sourceUrl;
  if (data.application?.url) {
    url = data.application.url;
  } else if (data.url) {
    url = data.url;
  }

  // Extract location
  let location: string | undefined;
  if (data.jobLocation) {
    if (Array.isArray(data.jobLocation) && data.jobLocation.length > 0) {
      const firstLoc = data.jobLocation[0];
      if (typeof firstLoc === 'object' && firstLoc !== null) {
        const locObj = firstLoc as Record<string, unknown>;
        location = (locObj.address as Record<string, unknown>)?.addressLocality as string | undefined;
      }
    } else if (typeof data.jobLocation === 'object' && data.jobLocation !== null) {
      const locObj = data.jobLocation as Record<string, unknown>;
      location = (locObj.address as Record<string, unknown>)?.addressLocality as string | undefined;
    }
  } else if (data.location) {
    location = data.location;
  }

  // Parse salary
  const salaryMin = parseSalary(data.estimatedSalary?.minValue);
  const salaryMax = parseSalary(data.estimatedSalary?.maxValue);

  return {
    title: data.title,
    company,
    description: data.description || data.jobDescription,
    url,
    location,
    remote: detectRemote(location || ''),
    job_type: data.employmentType,
    salary_min: salaryMin,
    salary_max: salaryMax,
  };
}

export function normalizeBatch(rawData: RawData[], sourceUrl: string): JobPosting[] {
  const normalized: JobPosting[] = [];

  for (const item of rawData) {
    try {
      const job = normalizeJobPosting(item, sourceUrl);
      if (job) {
        normalized.push(job);
      }
    } catch {
      continue;
    }
  }

  return normalized;
}

export type { JobPosting };
